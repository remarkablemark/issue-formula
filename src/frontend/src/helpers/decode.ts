export function decode(text: string) {
  return text.replaceAll('&gt;', '>').replaceAll('&lt;', '<');
}
