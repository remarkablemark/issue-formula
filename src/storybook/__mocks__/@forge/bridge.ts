import { fn } from '@storybook/test';

export const requestJira = fn();

export const view = {
  submit: fn(),
  close: fn(),
  refresh: fn(),
  createHistory: fn(),
  getContext: fn(),
  theme: {
    enable: fn(),
  },
};
