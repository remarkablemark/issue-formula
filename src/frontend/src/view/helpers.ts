import { Parser } from 'expr-eval';

import { log } from '../helpers';
import type { FormValues, Issue } from '../types';

/**
 * Transforms form values and issues into labels and values.
 *
 * @param formValues - Form values.
 * @param issues - Issues.
 * @returns - Array of label and value.
 */
export function transform(formValues: FormValues, issues: Issue[]) {
  const variables = getVariables(formValues, issues);

  return formValues.label.map((label, index) => {
    let value = '';
    const formula = formValues.formula[index];
    const decimal = Number(formValues.decimal[index]);

    try {
      value = Parser.evaluate(formula, variables).toFixed(decimal);
    } catch (error) {
      log.error('expr-eval:', error);
    }

    const prefix = formValues.prefix[index] || '';
    const suffix = formValues.suffix[index] || '';
    value = `${prefix}${value}${suffix}`;

    return {
      label,
      value,
    };
  });
}

/**
 * Gets variables from form values and issues.
 *
 * @param formValues - Form values.
 * @param issues - Issues.
 * @returns - Variables.
 */
export function getVariables(formValues: FormValues, issues: Issue[]) {
  return formValues.variable.reduce(
    (accumulator: Record<string, number>, variable, index) => {
      let value = NaN;
      const issue = issues[index];

      if (!issue) {
        return accumulator;
      }

      switch (formValues.function[index].value) {
        case 'COUNT':
          value = issues[index].total;
          break;
      }

      accumulator[variable] = value;
      return accumulator;
    },
    {},
  );
}
