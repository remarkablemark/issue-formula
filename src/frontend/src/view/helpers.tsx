import { Parser } from 'expr-eval';

import { log } from '../helpers';
import { FormValues, Issue } from '../types';

export function transform(formValues: FormValues, issues: Issue[]) {
  const variables = getVariables(formValues, issues);

  return formValues.label.map((label, index) => {
    let value = '';
    const formula = formValues.formula[index];
    const decimal = Number(formValues.decimal[index]);

    try {
      value = Parser.evaluate(formula, variables).toFixed(decimal);
    } catch (error) {
      log.error('expr-eval:', error);
    }

    const prefix = formValues.prefix[index] || '';
    const suffix = formValues.suffix[index] || '';
    value = `${prefix}${value}${suffix}`;

    return {
      label,
      value,
    };
  });
}

export function getVariables(formValues: FormValues, issues: Issue[]) {
  return formValues.variable.reduce(
    (accumulator: Record<string, number>, variable, index) => {
      let value = NaN;
      const issue = issues[index];

      if (!issue) {
        return accumulator;
      }

      switch (formValues.function[index].value) {
        case 'COUNT':
          value = issues[index].total;
          break;
      }

      accumulator[variable] = value;
      return accumulator;
    },
    {},
  );
}
