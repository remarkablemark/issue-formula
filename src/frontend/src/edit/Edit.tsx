import ButtonGroup from '@atlaskit/button/button-group';
import LoadingButton from '@atlaskit/button/loading-button';
import Button from '@atlaskit/button/new';
import Form, { FormFooter } from '@atlaskit/form';
import { view } from '@forge/bridge';

import { log } from '../helpers';
import { useFormValues, useGadgetConfiguration } from '../hooks';
import type { FormValues } from '../types';
import Formulas from './formulas';
import Variables from './variables';

export default function Edit() {
  const { formValues: oldFormValues } = useGadgetConfiguration();
  const formValues = useFormValues();

  return (
    <Form<FormValues>
      onSubmit={() => {
        import.meta.env.DEV && log.info('submit:', formValues);
        view.submit(formValues);
      }}
    >
      {({ formProps, submitting }) => (
        <form {...formProps}>
          <Variables />
          <Formulas />

          <FormFooter align="start">
            <ButtonGroup label="Form submit options">
              <LoadingButton
                type="submit"
                appearance="primary"
                isLoading={submitting}
              >
                Save
              </LoadingButton>

              <Button
                onClick={() => {
                  import.meta.env.DEV && log.info('cancel');
                  formValues && view.submit(oldFormValues);
                }}
              >
                Cancel
              </Button>
            </ButtonGroup>
          </FormFooter>
        </form>
      )}
    </Form>
  );
}
