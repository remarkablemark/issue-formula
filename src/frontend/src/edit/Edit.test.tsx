import { view } from '@forge/bridge';
import { render, screen, waitFor } from '@testing-library/react';

import { changeLabelValue, clickButton } from '../../test/helpers';
import type { FormValues, FullContext } from '../types';
import Edit from './Edit';

const mockedView = jest.mocked(view);
let consoleErrorSpy: jest.SpyInstance;

beforeAll(() => {
  // suppress React deprecated Context API error
  consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
});

afterAll(() => {
  consoleErrorSpy.mockRestore();
});

describe('without data', () => {
  beforeEach(async () => {
    mockedView.getContext.mockResolvedValueOnce({
      extension: { gadgetConfiguration: undefined },
    } as unknown as FullContext);

    render(<Edit />);
    await waitFor(() => {
      expect(screen.queryAllByRole('textbox')).toHaveLength(0);
    });
  });

  it.each(['Variables', 'Formulas'])('renders heading %s', (name) => {
    expect(screen.getByRole('heading', { level: 3, name })).toBeInTheDocument();
  });

  it.each(['Save', 'Cancel'])('renders button %s', (name) => {
    expect(screen.getByRole('button', { name })).toBeInTheDocument();
  });

  it('submits variable and formula', () => {
    clickButton('Add variable');
    [
      ['Variable', 'a'],
      ['JQL', 'created >= -30d'],
    ].forEach(([label, value]) => changeLabelValue(label, value));

    clickButton('Add formula');
    [
      ['Math Formula', 'a / 100'],
      ['Label', 'monthly issues'],
      ['Decimals', '2'],
      ['Prefix', '#'],
      ['Suffix', '%'],
    ].forEach(([label, value]) => changeLabelValue(label, value));

    clickButton('Save');
    expect(mockedView.submit).toHaveBeenCalledTimes(1);
    expect(mockedView.submit.mock.calls[0][0]).toMatchSnapshot();
  });

  it('cancels edit', () => {
    clickButton('Add variable');
    changeLabelValue('Variable', 'a');

    clickButton('Cancel');
    expect(mockedView.submit).toHaveBeenCalledTimes(1);
    expect(mockedView.submit).toHaveBeenCalledWith(undefined);
  });
});

describe('with data', () => {
  const formValues: FormValues = {
    decimal: ['2'],
    formula: ['a / 100'],
    function: [{ label: 'COUNT', value: 'COUNT' }],
    jql: ['created >= -30d'],
    label: ['monthly issues'],
    prefix: ['#'],
    suffix: ['%'],
    variable: ['a'],
  };

  beforeEach(async () => {
    mockedView.getContext.mockResolvedValueOnce({
      extension: { gadgetConfiguration: formValues },
    } as unknown as FullContext);

    render(<Edit />);
    await waitFor(() => {
      expect(screen.queryAllByRole('textbox')).toHaveLength(6);
    });
  });

  it('renders inputs', () => {
    Object.values(formValues).forEach(([value]) => {
      if (typeof value !== 'string') {
        value = value.value;
      }
      expect(screen.getByDisplayValue(value)).toBeInTheDocument();
    });
  });

  it('submits edited variable and formula', () => {
    [
      ['Variable', 'b'],
      ['JQL', 'created >= -7d'],
      ['Math Formula', 'b / 100'],
      ['Label', 'weekly issues'],
      ['Decimals', '1'],
      ['Prefix', '~'],
      ['Suffix', '*'],
    ].forEach(([label, value]) => changeLabelValue(label, value));

    clickButton('Save');
    expect(mockedView.submit).toHaveBeenCalledTimes(1);
    expect(mockedView.submit.mock.calls[0][0]).toMatchSnapshot();
  });

  it('cancels edit', () => {
    changeLabelValue('Variable', 'b');
    clickButton('Cancel');
    expect(mockedView.submit).toHaveBeenCalledTimes(1);
    expect(mockedView.submit).toHaveBeenCalledWith(formValues);
  });
});
