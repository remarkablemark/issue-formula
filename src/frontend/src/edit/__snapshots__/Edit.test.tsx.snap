// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with data submits edited variable and formula 1`] = `
{
  "decimal": [
    "1",
  ],
  "formula": [
    "b / 100",
  ],
  "function": [
    {
      "label": "COUNT",
      "value": "COUNT",
    },
  ],
  "jql": [
    "created >= -7d",
  ],
  "label": [
    "weekly issues",
  ],
  "prefix": [
    "~",
  ],
  "suffix": [
    "*",
  ],
  "variable": [
    "b",
  ],
}
`;

exports[`without data submits variable and formula 1`] = `
{
  "decimal": [
    "2",
  ],
  "formula": [
    "a / 100",
  ],
  "function": [
    {
      "label": "COUNT",
      "value": "COUNT",
    },
  ],
  "jql": [
    "created >= -30d",
  ],
  "label": [
    "monthly issues",
  ],
  "prefix": [
    "#",
  ],
  "suffix": [
    "%",
  ],
  "variable": [
    "a",
  ],
}
`;
